src/main/scala/is/hail/expr/types/TBaseStruct.scala

```diff
@@ -0,0 +1,216 @@
+package is.hail.expr.types
+
+import is.hail.annotations._
+import is.hail.asm4s.{Code, _}
+import is.hail.check.Gen
+import is.hail.utils._
+import org.apache.spark.sql.Row
+import org.json4s.jackson.JsonMethods
+
+import scala.reflect.{ClassTag, classTag}
+
+object TBaseStruct {
+  def getOrdering(types: Array[Type]): ExtendedOrdering =
+    ExtendedOrdering.rowOrdering(types.map(_.ordering))
+
+  def getMissingness(types: Array[Type], missingIdx: Array[Int]): Int = {
+    assert(missingIdx.length == types.length)
+    var i = 0
+    types.zipWithIndex.foreach { case (t, idx) =>
+      missingIdx(idx) = i
+      if (!t.required)
+        i += 1
+    }
+    i
+  }
+
+  def getByteSizeAndOffsets(types: Array[Type], nMissingBytes: Long, byteOffsets: Array[Long]): Long = {
+    assert(byteOffsets.length == types.length)
+    val bp = new BytePacker()
+
+    var offset: Long = nMissingBytes
+    types.zipWithIndex.foreach { case (t, i) =>
+      val fSize = t.byteSize
+      val fAlignment = t.alignment
+
+      bp.getSpace(fSize, fAlignment) match {
+        case Some(start) =>
+          byteOffsets(i) = start
+        case None =>
+          val mod = offset % fAlignment
+          if (mod != 0) {
+            val shift = fAlignment - mod
+            bp.insertSpace(shift, offset)
+            offset += (fAlignment - mod)
+          }
+          byteOffsets(i) = offset
+          offset += fSize
+      }
+    }
+    offset
+  }
+
+  def alignment(types: Array[Type]): Long = {
+    if (types.isEmpty)
+      1
+    else
+      types.map(_.alignment).max
+  }
+}
+
+abstract class TBaseStruct extends Type {
+  def types: Array[Type]
+
+  def fields: IndexedSeq[Field]
+  
+  def fieldRequired: Array[Boolean]
+
+  override def children: Seq[Type] = types
+
+  def size: Int
+
+  def _toString: String = {
```

Comment:
everything below here was just copied from TStruct, right? (I don't have to read it closely?)

Reply:
correct.